### ISSUE TITLE
'use autodie' causes Perl 5 build process to gag on podlators

### ISSUE THREAD

**Issue by jkeenan** [rra/podlators#32]
We are having two problems "synching" `podlators-v6.0.0` into the Perl core distribution.

The first problem is "on us" in the sense that it appears that our synching program `Porting/sync-with-cpan` is not yet adapted to one or both of "version declarations in the `package` statement" and "3-part semantic versioning."  We'll handle this in https://github.com/Perl/perl5/issues/22395.

The second problem lies in the fact that `podlators-v6.0.0`, unlike earlier versions of the library, makes extensive use of `use autodie`.  As it stands now, not all of `autodie`'s prerequisites are found in `@INC` at the point the build process enters `cpan/podlators/`.  You can reproduce this problem as follows:

1. Start with a clean checkout of Perl 5's main branch, "blead":
```
$ git describe
v5.41.1-35-gafff10bbde
$ git status
On branch blead
Your branch is up to date with 'origin/blead'.

nothing to commit, working tree clean
```
2. Run `./Configure` in that checkout; do not run `make`.
```
$ sh ./Configure -des -Dusedevel
```
3. Download a tarball of podlators-v6.0.0 from CPAN:
```
$ ls -ltr ~/Downloads | tail -n 1
-rw-rw-r-- 1 jkeenan jkeenan    158110 Jul 11 19:43 podlators-v6.0.0.tar.gz
```
4. Back in your checkout, first edit `Porting/Maintainers.pl` to remove the `CUSTOMIZED` element from the entry for `cpan/podlators`:
```
$ vi Porting/Maintainers.pl
[# edit and save such that we get this difference]

$ git diff
diff --git a/Porting/Maintainers.pl b/Porting/Maintainers.pl
index f90455ce23..867df3c076 100755
--- a/Porting/Maintainers.pl
+++ b/Porting/Maintainers.pl
@@ -988,15 +988,6 @@ our %Modules = (
             qr{^\.github/dependabot\.yml},
             qr{^\.github/workflows/build\.yaml},
         ],
-        'CUSTOMIZED'   => [
-            # https://github.com/rra/podlators/pull/28
-            'lib/Pod/Man.pm',
-            'lib/Pod/ParseLink.pm',
-            'lib/Pod/Text.pm',
-            'lib/Pod/Text/Color.pm',
-            'lib/Pod/Text/Overstrike.pm',
-            'lib/Pod/Text/Termcap.pm'
-        ],
         'MAP' => {
             ''                 => 'cpan/podlators/',
             # this file lives outside the cpan/ directory
```
5. As a safety precaution, save this change in a new branch.
```
$  git checkout -b xyz
Switched to a new branch 'xyz'

$ git status
On branch xyz
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   Porting/Maintainers.pl

no changes added to commit (use "git add" and/or "git commit -a")

$ git add Porting/Maintainers.pl 

$ git commit -m "Remove CUSTOMIZED entry for podlators"
[xyz b8b1c73c82] Remove CUSTOMIZED entry for podlators
 1 file changed, 9 deletions(-)
```
6. Now run the synching program with the `--tarball` option.  There will be a couple of points where you are prompted to hit Return; do so as long as the program continues to run.  (Probably 3 Returns before program fails.)
```
$ perl Porting/sync-with-cpan podlators --tarball=/home/jkeenan/Downloads/podlators-v6.0.0.tar.gz
...
Starting `make test-prep` and saving its output to 'make.log' ...
Finished `make test-prep`
Running `make` failed, see 'make.log' for more details
```
7. Inspect `make.log`.  Of note:

* Around line 76, you should see:
```
./miniperl -Ilib make_ext.pl cpan/autodie/pm_to_blib  MAKE="make" LIBPERL_A=libperl.a
Running pm_to_blib for cpan/autodie directly
```
So we ran the extension-building program, `make_ext.pl` via `miniperl` and were able to build the `autodie` module, which is used extensively in podlators-v6.0.0, but not in the older podlators-5.0.1.

* Later, at around 308, we try to build the podlators extension itself:
```
...
./miniperl -Ilib make_ext.pl cpan/podlators/pm_to_blib  MAKE="make" LIBPERL_A=libperl.a
Can't locate Tie/RefHash.pm in @INC (you may need to install the Tie::RefHash module) (@INC entries checked: /home/jkeenan/gitwork/perl2/cpan/AutoLoader/lib /home/jkeenan/gitwork/perl2/dist/Carp/lib /home/jkeenan/gitwork/perl2/dist/PathTools /home/jkeenan/gitwork/perl2/dist/PathTools/lib /home/jkeenan/gitwork/perl2/cpan/ExtUtils-Install/lib /home/jkeenan/gitwork/perl2/cpan/ExtUtils-MakeMaker/lib /home/jkeenan/gitwork/perl2/cpan/ExtUtils-Manifest/lib /home/jkeenan/gitwork/perl2/cpan/File-Path/lib /home/jkeenan/gitwork/perl2/ext/re /home/jkeenan/gitwork/perl2/dist/Term-ReadLine/lib /home/jkeenan/gitwork/perl2/dist/Exporter/lib /home/jkeenan/gitwork/perl2/ext/File-Find/lib /home/jkeenan/gitwork/perl2/cpan/Text-Tabs/lib /home/jkeenan/gitwork/perl2/dist/constant/lib /home/jkeenan/gitwork/perl2/cpan/version/lib /home/jkeenan/gitwork/perl2/cpan/Getopt-Long/lib /home/jkeenan/gitwork/perl2/cpan/Text-ParseWords/lib /home/jkeenan/gitwork/perl2/cpan/ExtUtils-PL2Bat/lib /home/jkeenan/gitwork/perl2/lib .) at /home/jkeenan/gitwork/perl2/lib/Fatal.pm line 9.
BEGIN failed--compilation aborted at /home/jkeenan/gitwork/perl2/lib/Fatal.pm line 9.
Compilation failed in require at /home/jkeenan/gitwork/perl2/lib/parent.pm line 16.
BEGIN failed--compilation aborted at /home/jkeenan/gitwork/perl2/lib/autodie.pm line 6.
Compilation failed in require at Makefile.PL line 16.
BEGIN failed--compilation aborted at Makefile.PL line 16.
Unsuccessful Makefile.PL(cpan/podlators): code=512 at make_ext.pl line 532.
make: *** [makefile:599: cpan/podlators/pm_to_blib] Error 2
```
Here I confess that I'm not the best person at reading a log file backwards for debugging.  But I know that `make_ext.pl` takes one approach when an "outside" library (one found in `cpan`, `dist` or `ext` in the Perl core distribution) has its own `Makefile.PL` and a different approach when the distro does not and relies upon the core to generate a Makefile.  `cpan/podlators/` has its own `Makefile.PL`, so `miniperl` calls `make_ext.pl` on `cpan/podlators/Makefile.PL` and compilation fails at its line 16.
```
 15 use 5.012;
 16 use autodie;
 17 use warnings;
```
`autodie`, not surprisingly, has `Fatal` as a parent:
```
  1 package autodie;
  2 use 5.008;
  3 use strict;
  4 use warnings;
  5 
  6 use parent qw(Fatal);
  7 our $VERSION;
```
`parent` has an `import` method that dies if a required file cannot be found:
```
  6 sub import {
  7     my $class = shift;
  8 
  9     my $inheritor = caller(0);
 10 
 11     if ( @_ and $_[0] eq '-norequire' ) {
 12         shift @_;
 13     } else {
 14         for ( my @filename = @_ ) {
 15             s{::|'}{/}g;
 16             require "$_.pm"; # dies if the file is not found
 17         }
```
And what turns out to be missing at line 9 of `lib/Fatal.pm` is:
```
   1 package Fatal; 
   2 
   3 # ABSTRACT: Replace functions with equivalents which succeed or die
   4 
   5 use 5.008;  # 5.8.x needed for autodie 
   6 use Carp;
   7 use strict;
   8 use warnings;
   9 use Tie::RefHash;   # To cache subroutine refs
```
... `Tie::RefHash`, a library shipped with core that is so obscure than until today I had never heard of it notwithstanding having worked on core for a dozen years!

Now, what would I have expected to see in a log of the output of `make test_prep` in, say, blead.  Well, if I ran verbosely (`make --debug=v`), I would get something like this:
```
 315 ./miniperl -Ilib make_ext.pl cpan/podlators/pm_to_blib  MAKE="make" LIBPERL_A=libperl.a
 316 Generating a Unix-style Makefile
 317 Writing Makefile for Pod
 318 make[1]: Entering directory '/home/jkeenan/gitwork/perl/cpan/podlators'
 319 "../../miniperl" "-I../../lib" "-I../../lib" scripts/pod2man.PL scripts/pod2man
 320 Extracting pod2man (with variable substitutions)
 321 "../../miniperl" "-I../../lib" "-I../../lib" scripts/pod2text.PL scripts/pod2text
 322 Extracting pod2text (with variable substitutions)
 323 cp scripts/pod2man blib/script/pod2man
 324 /home/jkeenan/gitwork/perl/cpan/podlators/../../miniperl "-I../../lib" -MExtUtils::MY -e 'MY->fixin     (shift)' -- blib/script/pod2man
 325 cp scripts/pod2text blib/script/pod2text
 326 /home/jkeenan/gitwork/perl/cpan/podlators/../../miniperl "-I../../lib" -MExtUtils::MY -e 'MY->fixin     (shift)' -- blib/script/pod2text
 327 Manifying 2 pod documents
 328 make[1]: Leaving directory '/home/jkeenan/gitwork/perl/cpan/podlators'
```
... and then farther down, we would get to `Tie::RefHash`:
```
 406 ./miniperl -Ilib make_ext.pl cpan/Tie-RefHash/pm_to_blib  MAKE="make" LIBPERL_A=libperl.a
 407 Running pm_to_blib for cpan/Tie-RefHash directly
```
The implication is that once we start using `use autodie;` in podlator's `Makefile.PL` (and probably also once we use it in other files in the distribution (`.pm`, `.PL`, etc.)), then the Perl build process expects `autodie` **and all of its prerequisites, including Tie::RefHash,** to already be "in position", *i.e.,* locatable in `@INC`.  I do not currently know how to "fix" the build process to guarantee that all of `autodie`'s prequisites are built before any library whose `Makefile.PL` `use`s `autodie` is run by `miniperl make_ext.pl`.

What I *do* know, however, is that if I clone the `podlators-v6.0.0` tarball, edit it so as to (a) comment out all instances of `use autodie;`; and (b) change the version number to an arbitrary `6.0.99`; and call:
```
$ perl Porting/sync-with-cpan podlators --tarball=/home/jkeenan/tmp/podlators-v6.0.99.tar.gz
```
... then this tarball is correctly synched into the Perl core distribution.

It appears that no "outside" module shipped with the core distribution itself calls `use autodie;` itself.  This may be because `miniperl` itself is not equipped to deal with `autodie` or its prerequisites, as this obscure comment in the core suggests:
```
$ grep -n 'use autodie' lib/ExtUtils/Miniperl.pm 
10:# blead will run this with miniperl, hence we can't use autodie or File::Temp
```
Now, I suspect that once the Perl build process gets set to all `make test`, all the necessary libraries are "in position" in `@INC` such that you could say `use autodie;` in the podlators test suite.  But I ask you to reconsider your use of `use autodie;` in `Makefile.PL`, all the `lib/*/*.pm` files and the executable-generator `*.PL` files.

Thank you very much.
Jim Keenan

**Comment by rra** [comment]
It sounds like maybe the only place it can't be used where it's used now is in `Makefile.PL` and `pod2man.PL`. Since you have a test environment already built, could you try deleting the `use autodie` line from just those two files and see if everything works? If so, I'll put out a quick 6.0.1 release with that fix.

**Comment by jkeenan** [comment]
On 7/11/24 21:15, Russ Allbery wrote:
> It sounds like maybe the only place it can't be used where it's used now 
> is in |Makefile.PL| and |pod2man.PL|. Since you have a test environment 
> already built, could you try deleting the |use autodie| line from just 
> those two files and see if everything works? If so, I'll put out a quick 
> 6.0.1 release with that fix.
> 

It's late here, so I'll try that tomorrow, perhaps as a pull request.

**Comment by rra** [comment]
Oh, yeah, whenever you get a chance, and also feel free to ask me to test it myself.

Don't worry about a PR unless you want to. I will end up adding some comments and whatnot too. I should have figured that using `autodie` in the build system was a bit aggressive, and just want to be sure that there aren't any other lingering issues before I put out a new release.

**Comment by rra** [comment]
Thanks, hopefully fixed in v6.0.1.
